### ISSUE TITLE
Is there a WYSIWYG tool to edit MAML files

### ISSUE THREAD

**Issue by wmjordan** [EWSoftware/SHFB#508]
Instead of messing with MAML file tags, is there a WYSIWYG tool to edit MAML files?

**Comment by EWSoftware** [comment]
There aren't any available that I'm aware of.  Dave Sexton started to [create one](http://davesexton.com/blog/post/maml-editor_2D00_progress.aspx) but as far as I know it was abandoned and the source code was never released.

@RxDave Not sure if you'll see this but, if so, would you consider open sourcing what you did complete of your MAML editor so that I or someone else can have a go at finishing it?

**Comment by RxDave** [comment]
That should be doable. I don't remember the state in which I left it, but I'll take a look after work today.

**Comment by sharwell** [comment]
👍

**Comment by RxDave** [comment]
I don't think it worked reasonably well, which is why I never followed through with it :D But let's see.

**Comment by RxDave** [comment]
Here ya go: 
https://github.com/RxDave/XMLGel

I'm interested to see if you can do something with this.

**Comment by EWSoftware** [comment]
Thanks, I appreciate it.  I'll see what I can do.

**Comment by wmjordan** [comment]
Thank you! @RxDave and @EWSoftware!
I am looking forward to the reincarnation of the resurrected project.

**Comment by wmjordan** [comment]
@EWSoftware Maybe there is an alternative to build a visual MAML editor.
Since HTML editors are relatively easier to find, maybe it is possible to implement the WYSIWYG HTML->MAML editor in an XSLT-supported way.
1. Program some XSLT documents,
2. Author help files with HTML editors, which generate documents with simple markups, such as &lt;p>, &lt;h1>, &lt;a>, etc, 
3. Use XSLT to transform HTML documents (with the help of some existing HTML readers, for instance, https://github.com/lovettchris/SgmlReader) into the MAML format.
Or even simpler, use an HTML reader to extract the &lt;body> content, and inject it into HTML templates.

**Comment by EWSoftware** [comment]
MAML documents are much more structured than HTML and there are several different types.  Each document type can vary with regard to required and optional elements, subsections can be nested within other sections, and there are many places where attributes define how a particular element is rendered (i.e. code block language and title, alert types, not to mention addressable elements like sections that can be used in link targets.  I don't think that kind of stuff can be easily put into an HTML editor and round tripped between formats.

Also, something like the glossary document type wouldn't work well being edited in a WYSIWYG format.  Those would be better off with some form of customized UI to define the glossary terms and edit their definitions.

The point of using MAML is that you define the structure of the content, not the layout or style.  The presentation style handles that part.  It also allows you to generate the MAML content in different output formats (HTML, markdown, or Open XML for example).  Using a simplistic approach where all you use is HTML templates loses all of that functionality.

**Comment by wmjordan** [comment]
I found an HTML to MAML Converter tool in the extra folder.
I thought it was possible to utilize that tool to easier the process of MAML authoring.
However, I could not figure out how to make a link which point to the API documentation part.

**Comment by EWSoftware** [comment]
The intended use for the HTML to MAML converter is to convert old HTML content to MAML content in a one time conversion.  As noted in the [help for it](http://ewsoftware.github.io/HtmlToMamlDocs/html/a5f36b5f-5551-43c4-a938-c1451571b97f.htm), there can be several tasks to handle in the converted content so it's not for general use on a regular basis.

**Comment by PaulStSmith** [comment]
Thanks for the clarification, Eric — that makes perfect sense.

I can see how a Markdown→MAML hybrid approach would be far more sustainable than maintaining a custom WYSIWYG surface. The sample you shared paints a clear picture of how this could evolve incrementally without needing to rewrite the editor infrastructure.

That said, I’m still curious whether there’s interest in a minimal visual helper — not a full editor, but perhaps a pane that renders Markdown/MAML live for quick preview and structural validation (e.g., sections, IDs, outlines). Something akin to how Markdown editors show a split view.

If that’s in line with what the community would find useful, I could explore adapting the XMLGel fork or a WPF-based alternative purely for live rendering and editing assistance — not as a replacement editor.

Would that be something worth experimenting with, or do you think it’s better to focus efforts directly on the Markdown→MAML converter first?

**Comment by EWSoftware** [comment]
I've implemented support for using Markdown rather than MAML for conceptual topics.  Using Markdig it was pretty straightforward to create a topic renderer.  The majority of the time spent on the updates was for the conversion feature and updating the topic previewer to support the HTML elements.  I still need to do some more testing by converting the help for my own projects and to update the SHFB help to reflects all of the changes.
